File: Codes/139_wordBreak/139.单词拆分.py
'''
Description: 
Autor: HTmonster
Date: 2022-03-22 10:20:35
'''
#
# @lc app=leetcode.cn id=139 lang=python3
#
# [139] 单词拆分
#

# @lc code=start
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        
        # 使用备忘录
        mem={} #idx:bool 位置及s[idx:]的匹配情况

        # 子问题
        def __word_break(idx):
            # 退出条件
            if idx>len(s)-1: return True

            # 直接结果返回
            if idx in mem.keys():
                return mem[idx]
            # 遍历每个词
            rst=False
            for w in wordDict:
                # 问题在于 拆不拆
                if s[idx:].startswith(w):
                    rst= rst or __word_break(idx+len(w))
                    if rst: break
            mem[idx]=rst
            return rst

        return __word_break(0)
# @lc code=end

File: Codes/139_wordBreak/139.单词拆分.go
/*
 * @Description:
 * @Autor: HTmonster
 * @Date: 2022-03-22 10:42:37
 */
/*
 * @lc app=leetcode.cn id=139 lang=golang
 *
 * [139] 单词拆分
 */

// @lc code=start
func wordBreak(s string, wordDict []string) bool {

	// 备忘录
	mem := make(map[int]bool) //位置:是否能解决

	var subfunc func(int) bool //子函数
	subfunc = func(idx int) bool {
		// 退出条件
		if idx > len(s)-1 {
			return true
		}
		// 查备忘录 存在则直接返回结果
		if r, ok := mem[idx]; ok {
			return r
		}

		// 遍历字典
		flag := false
		for i := range wordDict {
			if strings.HasPrefix(s[idx:], wordDict[i]) {
				flag = flag || subfunc(idx+len(wordDict[i]))
				if flag {
					break
				}
			}
		}
		// 添加备忘录
		mem[idx] = flag

		return flag
	}

	return subfunc(0)

}

// @lc code=end

